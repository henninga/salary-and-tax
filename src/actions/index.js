import Cookie from 'js-cookie';
import { updateIntl } from 'react-intl-redux';
import { submit } from 'redux-form';
import { getMessages, toggleLocale } from '../translation';


export const FORM_CHANGE = 'FORM_CHANGE';
export const CHANGE_LANGUAGE = 'CHANGE_LANGUAGE';
export const UPDATED_LANGUAGE = 'UPDATED_LANGUAGE';

export const formChange = (values) => ({type: FORM_CHANGE, ...values});
export const inputHasChanged = (values) => dispatch => dispatch(formChange(values));

export const changeLanguage = (locale) => ({type: CHANGE_LANGUAGE, locale});
export const updatedLanguage = (locale) => ({type: UPDATED_LANGUAGE, locale});

export const switchLanguage = (currentLocale) => dispatch => {
  dispatch(changeLanguage(currentLocale));
  const locale = toggleLocale(currentLocale);
  Cookie.set('locale', locale);
  dispatch(updateIntl({locale, messages: getMessages(locale)}));
  dispatch(updatedLanguage(locale));
  dispatch(submit({form: 'parameters'}))
};
