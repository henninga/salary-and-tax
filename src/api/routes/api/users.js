import express from 'express';
import getUsers from '../../cvpartner/getUsers';

const router = express.Router();

router.get('/', (req, res) => {
    getUsers()
        .then(users => {
            res.status(200).json(users);
        })
        .catch(({ error: { status, error } }) => {
            res.status(status).json({ error });
        });
});

module.exports = router;
