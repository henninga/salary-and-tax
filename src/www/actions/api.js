import fetch from 'cross-fetch';


export const RECEIVE_USERS = 'RECEIVE_USERS';
export const REQUEST_USERS = 'REQUEST_USERS';
export const USER_IS_SELECTED = 'USER_IS_SELECTED';

export const REQUEST_CV = 'REQUEST_CV';
export const RECEIVE_CV = 'RECEIVE_CV';

export const FORM_CHANGE = 'FORM_CHANGE';

export const requestUsers = () => ({
    type: REQUEST_USERS,
});

export const receiveUsers = (list) => ({
    type: RECEIVE_USERS,
    list,
});

export const requestCv = (userId, cvId, name) => ({
    type: REQUEST_CV,
    userId,
    cvId,
    name,
});

export const receiveCv = (cv) => ({
    type: RECEIVE_CV,
    cv,
});

export const formChange = (values) => ({
  type: FORM_CHANGE,
  ...values,
});

export const fetchUsers = () => {
    return dispatch => {
        dispatch(requestUsers());
        return fetch('http://localhost:3000/api/users')
            .then(response => response.json())
            .then(list => dispatch(receiveUsers(list)))
            .then((users) => {
                if (users.list.length === 1) { dispatch(selectUser(users.list[0])) }
                return users;
            });
    }
};

export const selectUser = ({userId, cvId, name}) => {
    return dispatch => {
        dispatch(requestCv(userId, cvId, name));
        return fetch(`http://localhost:3000/api/cv/${userId}/${cvId}`)
            .then(response => response.json())
            .then(json => dispatch(receiveCv(json)))
    }
};

export const inputHasChanged = (values) => {
  return dispatch => {
    dispatch(formChange(values));
  }
};
