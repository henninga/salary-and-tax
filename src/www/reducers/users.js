import { RECEIVE_USERS, REQUEST_USERS } from '../actions';


const getSelectedUser = (list) => {
  return list && list.length === 1 ? {...list[0]} : {};
};

const initialList = [];

const initialUsersState = {
  isFetching: false,
  isLoaded: false,
  list: initialList,
  selectedUser: getSelectedUser(initialList),
};

const users = (state = initialUsersState, action = {}) => {
  switch (action.type) {
    case REQUEST_USERS:
      return Object.assign({}, state, {
        isFetching: true,
        isLoaded: false,
      });
    case RECEIVE_USERS:
      return Object.assign({}, state, {
        isFetching: false,
        isLoaded: true,
        list: action.list,
        selectedUser: getSelectedUser(action.list),
      });
    default:
      return state;
  }
};

export default users;
