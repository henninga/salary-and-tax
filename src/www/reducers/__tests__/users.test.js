import { RECEIVE_USERS, REQUEST_USERS } from '../../actions';
import users from '../users';

describe('users reducer', function () {
    it('should have default values that match snapshot', function () {
        expect(users()).toMatchSnapshot();
    });
    describe('action REQUEST_USERS', function () {
        it('should set isFetching to false', function () {
            expect(users(undefined, { type: 'IRRELEVANT' }).isFetching).toBeFalsy();
            expect(users(undefined, { type: REQUEST_USERS }).isFetching).toBeTruthy();
            expect(users({ isFetching: true }, { type: REQUEST_USERS }).isFetching).toBeTruthy();
            expect(users({ isFetching: false }, { type: REQUEST_USERS }).isFetching).toBeTruthy();
        });
        it('should only change isFetching', function () {
            const state = {
                some: 'state',
                which: 'should',
                be: 'kept',
            };
            expect(users(state, { type: REQUEST_USERS })).toEqual(expect.objectContaining(state));
        });
    });
    describe('action RECEIVE_USERS', function () {
        it('should set isFetching to false', function () {
            expect(users(undefined, { type: 'IRRELEVANT' }).isFetching).toBeFalsy();
            expect(users(undefined, { type: RECEIVE_USERS }).isFetching).toBeFalsy();
            expect(users({ isFetching: true }, { type: RECEIVE_USERS }).isFetching).toBeFalsy();
            expect(users({ isFetching: false }, { type: RECEIVE_USERS }).isFetching).toBeFalsy();
        });
        it('should add list of users', function () {
            expect(users(undefined,
                {
                    type: RECEIVE_USERS,
                    list: [
                        { user: '1' },
                        { user: '2' },
                    ],
                })).toMatchSnapshot();
        });
        it('should only change isFetching', function () {
            const state = {
                some: 'state',
                which: 'should',
                be: 'kept',
            };
            expect(users(state, { type: RECEIVE_USERS })).toEqual(expect.objectContaining(state));
        });
    });
});
