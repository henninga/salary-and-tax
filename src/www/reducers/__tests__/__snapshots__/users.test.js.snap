// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users reducer action RECEIVE_USERS should add list of users 1`] = `
Object {
  "isFetching": false,
  "isLoaded": true,
  "list": Array [
    Object {
      "user": "1",
    },
    Object {
      "user": "2",
    },
  ],
  "selectedUser": Object {},
}
`;

exports[`users reducer should have default values that match snapshot 1`] = `
Object {
  "isFetching": false,
  "isLoaded": false,
  "list": Array [],
  "selectedUser": Object {},
}
`;
